File: csharp/pack/packable/PackBuffer.cs
using System;


namespace pack.packable
{
    public class PackBuffer
    {
        internal byte[] hb;
        internal int position;
        internal int limit;

        public PackBuffer(byte[] array, int offset, int length)
        {
            hb = array;
            position = offset;
            limit = offset + length;
        }

        public PackBuffer(byte[] array) : this(array, 0, array.Length)
        {
        }

        internal bool HasRemaining()
        {
            return position < limit;
        }

        internal void CheckBound(int offset, int len)
        {
            if (offset + len > limit)
            {
                throw new IndexOutOfRangeException("buffer out of bound");
            }
        }

        public void WriteByte(byte x)
        {
            hb[position++] = x;
        }

        public byte ReadByte()
        {
            return hb[position++];
        }

        public static int GetVarint32Size(int value)
        {
            uint x = (uint)value;
            if (x <= 0x7f)
            {
                return 1;
            }
            else if (x <= 0x3fff)
            {
                return 2;
            }
            else if (x <= 0x1fffff)
            {
                return 3;
            }
            else if (x <= 0xfffffff)
            {
                return 4;
            }
            return 5;
        }

        public void WriteVarint32(int value)
        {
            uint x = (uint)value;
            while (x > 0x7f)
            {
                hb[position++] = (byte)((x & 0x7f) | 0x80);
                x >>= 7;
            }
            hb[position++] = (byte)x;
        }

        public int ReadVarint32()
        {
            uint x = hb[position++];
            if (x <= 0x7f) return (int)x;
            x = x & 0x7f | ((uint)hb[position++] << 7);
            if (x <= 0x3fff) return (int)x;
            x = (x & 0x3fff) | ((uint)hb[position++] << 14);
            if (x <= 0x1fffff) return (int)x;
            x = (x & 0x1fffff) | ((uint)hb[position++] << 21);
            if (x <= 0xfffffff) return (int)x;
            x = (x & 0xfffffff) | ((uint)hb[position++] << 28);
            return (int)x;
        }

        public void WriteBytes(byte[] bytes)
        {
            int count = bytes.Length;
            Buffer.BlockCopy(bytes, 0, hb, position, count);
            position += count;
        }

        public void ReadBytes(byte[] bytes)
        {
            int count = bytes.Length;
            Buffer.BlockCopy(hb, position, bytes, 0, count);
            position += count;
        }

        public void WriteShort(int i, short x)
        {
            hb[i++] = (byte)x;
            hb[i] = (byte)(x >> 8);
        }

        public void WriteShort(short x)
        {
            hb[position++] = (byte)x;
            hb[position++] = (byte)(x >> 8);
        }

        public short ReadShort()
        {
            return (short)(hb[position++] | (hb[position++] << 8));
        }

        public void WriteInt(int x)
        {
            hb[position++] = (byte)x;
            hb[position++] = (byte)(x >> 8);
            hb[position++] = (byte)(x >> 16);
            hb[position++] = (byte)(x >> 24);
        }

        public void WriteInt(int i, int x)
        {
            hb[i++] = (byte)x;
            hb[i++] = (byte)(x >> 8);
            hb[i++] = (byte)(x >> 16);
            hb[i] = (byte)(x >> 24);
        }

        public int ReadInt()
        {
            return hb[position++] |
                (hb[position++] << 8) |
                (hb[position++] << 16) |
                (hb[position++] << 24);
        }

        public void WriteLong(long x)
        {
            if (BitConverter.IsLittleEndian)
            {
                unsafe
                {
                    fixed (byte* pB = &hb[position])
                    {
                        position += 8;
                        *(long*)pB = x;
                    }
                }
            }
            else
            {
                hb[position++] = (byte)x;
                hb[position++] = (byte)(x >> 8);
                hb[position++] = (byte)(x >> 16);
                hb[position++] = (byte)(x >> 24);
                hb[position++] = (byte)(x >> 32);
                hb[position++] = (byte)(x >> 40);
                hb[position++] = (byte)(x >> 48);
                hb[position++] = (byte)(x >> 56);
            }
        }

        public long ReadLong(int i)
        {
            if (BitConverter.IsLittleEndian)
            {
                if (i + 8 > limit) throw new IndexOutOfRangeException("reading out of range");
                unsafe
                {
                    fixed (byte* pB = &hb[i])
                    {
                        return *(long*)pB;
                    }
                }
            }
            else
            {
                return hb[i++] |
                (((long)hb[i++]) << 8) |
                (((long)hb[i++]) << 16) |
                (((long)hb[i++]) << 24) |
                (((long)hb[i++]) << 32) |
                (((long)hb[i++]) << 40) |
                (((long)hb[i++]) << 48) |
                (((long)hb[i]) << 56);
            }

        }

        public long ReadLong()
        {
            long value = ReadLong(position);
            position += 8;
            return value;
        }

        public void WriteFloat(float x)
        {
            unsafe
            {
                WriteInt(*(int*)&x);
            }
        }

        public float ReadFloat()
        {
            int x = ReadInt();
            unsafe
            {
                return *(float*)&x;
            }
        }

        public void WriteDouble(double x)
        {
            unsafe
            {
                WriteLong(*(long*)&x);
            }
        }

        public double ReadDouble()
        {
            long x = ReadLong();
            unsafe
            {
                return *(double*)&x;
            }
        }
    }
}

File: objectivec/pack/PackBuffer.m

#import <Foundation/Foundation.h>
#import "PackBuffer.h"


@implementation PackBuffer

@synthesize hb;
@synthesize position;
@synthesize limit;

-(id)initWithBytes:(uint8_t*)array andOffset:(int)offset andLimit:(int)length {
    self = [super init];
    if(self != nil){
        hb = array;
        position = offset;
        limit = offset +length;
    }
    return self;
}

-(bool)hasRemaining{
    return position < limit;
}

-(void)checkBould:(int)offset andLen:(int)len{
    if(offset + len > limit){
        @throw([NSException exceptionWithName:@"CheckBouldFailed" reason:@"buffer out of bound" userInfo:nil]);
    }
}

-(void)throwOutOfBound:(NSString*)info atPos:(int)pos {
    NSString* msg = [[NSString alloc] initWithFormat:@"Reading buffer out of bound %@ pos: %d, limit:%d", info, pos, limit];
    @throw([NSException exceptionWithName:@"OutOfBound" reason:msg userInfo:nil]);
}

-(void)writeByte:(uint8) x {
    hb[position++] = x;
}

-(uint8)readByte:(int) i{
    if(i >= limit){
        [self throwOutOfBound:@"readByte at " atPos:i];
    }
    return hb[i];
}

-(uint8)readByte{
    if(position >= limit){
        [self throwOutOfBound:@"readByte" atPos:position];
    }
    return hb[position++];
}

-(void)writeShort:(int)i withValue:(short)x{
    hb[i] = x;
    hb[i+1] = x >> 8;
}

-(void)writeShort:(int)x{
    hb[position++] = x;
    hb[position++] = x >> 8;
}

-(short)readShort{
    if (position + 2 > limit) {
        [self throwOutOfBound:@"readShort" atPos:position];
    }
    short value = *((short *) (hb + position));
    position += 2;
    return value;
}

-(void)writeInt:(int)x{
    *((int *) (hb + position)) = x;
    position += 4;
}

-(void)writeInt:(int)i withValue:(int)x{
    *((int *) (hb + i)) = x;
}

-(int)readInt{
    if(position + 4 > limit){
        [self throwOutOfBound:@"readInt" atPos:position];
    }
    int t =  *((int *) (hb + position));
    position += 4;
    return t;
}


+(int)getVarint32Size:(int)x{
    uint32 t = x;
    if(t <= 0x7f){
        return 1;
    }else if(x <= 0x3fff){
        return 2;
    }else if(x <= 0x1fffff){
        return 3;
    }else if (x <= 0xfffffff) {
        return 4;
    }
    return 5;
}

-(void)writeVarintNegatie1{
    *((int *) (hb + position)) = 0xffffffff;
    hb[position+4] = 0xf;
    position += 5;
}

-(void)writeVarint32:(int)x{
    uint32 t = x;
    while (t > 0x7f) {
        hb[position++] = (uint8) ((t & 0x7f) | 0x80);
        t >>= 7;
    }
    hb[position++] = (uint8) t;
}

-(int)readVarint32{
    uint32 x = hb[position++];
    if (x <= 0x7f) goto end;
    x = (x & 0x7f) | (hb[position++] << 7);
    if (x <= 0x3fff) goto end;
    x = (x & 0x3fff) | (hb[position++] << 14);
    if (x <= 0x1fffff) goto end;
    x = (x & 0x1fffff) | (hb[position++] << 21);
    if (x <= 0xfffffff) goto end;
    x = (x & 0xfffffff) | (hb[position++] << 28);
end:
    if (position > limit){
        [self throwOutOfBound:@"readVarint32" atPos:position];
    }
    return x;
}

-(void)writeInt64:(int64_t)x{
    *((int64_t *) (hb + position)) = x;
    position += 8;
}

-(int64_t)readInt64:(int)i{
    if(i + 8 > limit){
        [self throwOutOfBound:@"readInt64 at " atPos:i];
    }
    return *((int64_t *) (hb + i));
}

-(int64_t)readInt64{
    if(position + 8 > limit){
        [self throwOutOfBound:@"readInt64  " atPos:position];
    }
    int64_t t = *((int64_t *) (hb + position));
    position += 8;
    return t;
}

-(void)writeFloat:(float)x{
    *((float *) (hb + position)) = x;
    position += 4;
}

-(float)readFloat{
    int t = [self readInt];
    return *((float *) (&t));
}

-(void)writeDouble:(double)x{
    *((double *) (hb + position)) = x;
     position += 8;
}

-(double)readDouble{
    int64_t t = [self readInt64];
    return *((double *) (&t));
}

-(void)writeBytes:(uint8*)src withSize:(int)n{
    if (n > 0) {
        memcpy(hb + position, src, n);
        position += n;
    }
}
-(void)readBytes:(uint8*) bytes withSize:(int)n{
    memcpy(bytes, hb + position, n);
    position += n;
}

@end
